### PR TITLE
Optimize group by

### PR DESCRIPTION
**Pull request by ghislainfourny**

### CODE REVIEW

**Comment by ghislainfourny** [comment]
Et voilà :) Can you try to measure and see if this is faster?

File: src/main/java/sparksoniq/spark/udf/GroupClauseCreateColumnsUDF.java [hunk 1]
```diff
@@ -71,9 +72,12 @@ public Row call(WrappedArray wrappedParameters) {
             // nulls and empty sequences have special grouping captured in the first grouping column
             // if non-null, non-empty-sequence value is given, the second column is used to group the input
             // indices are assigned to each value type for the first column
-            int emptySequenceGroupIndex = 1;         // by default, empty sequence is taken as first(=least)
-            int nullGroupIndex = 2;                  // null is the smallest value except empty sequence(default)
-            int valueGroupIndex = 3;                 // values are larger than null and empty sequence(default)
+            int emptySequenceGroupIndex = 1;
+            int nullGroupIndex = 2;
+            int trueGroupIndex = 3;
```
Comment (CanBerker) [inline src/main/java/sparksoniq/spark/udf/GroupClauseCreateColumnsUDF.java]: I think I would prefer booleanTrueGroupIndex and booleanFalseGroupIndex here for clarity. What do you think?

File: src/main/resources/test_files/runtime-spark/DataFrames/GroupbyClauseError1.iq [hunk 2]
```diff
@@ -1,4 +1,4 @@
-(:JIQS: ShouldCrash; ErrorCode="XPTY0004" :)
+(:JIQS: ShouldRun; Output="({ "guess" : "Bulgarian", "target" : "Albanian", "country" : 123, "choices" : [ "Albanian", "Bulgarian", "Russian", "Ukrainian" ], "sample" : "00b85faa8b878a14f8781be334deb137", "date" : "2013-08-19" }, { "guess" : "Bulgarian", "target" : "Albanian", "country" : "AU", "choices" : [ "Albanian", "Bulgarian", "Russian", "Ukrainian" ], "sample" : "00b85faa8b878a14f8781be334deb137", "date" : "2013-08-19" })" :)
```
Comment (CanBerker) [inline src/main/resources/test_files/runtime-spark/DataFrames/GroupbyClauseError1.iq]: I think we should rename the file to GroupbyClause8.iq as it's not expecting an error any more. Also extending this test with a few more diverse fields would be good for completeness

File: src/main/java/sparksoniq/spark/udf/GroupClauseCreateColumnsUDF.java [hunk 3]
```diff
@@ -84,37 +88,52 @@ public Row call(WrappedArray wrappedParameters) {
...
+                    _results.add(doubleGroupIndex);
+                    _results.add(null);
+                    _results.add(new Double(nextItem.getIntegerValue()));
+                } else if (nextItem.isDecimal()) {
+                    // any other atomic type
+                    _results.add(doubleGroupIndex);
+                    _results.add(null);
+                    _results.add(new Double(nextItem.getDecimalValue().doubleValue()));
+                } else if (nextItem.isDouble()) {
+                    // any other atomic type
```
Comment (CanBerker) [inline src/main/java/sparksoniq/spark/udf/GroupClauseCreateColumnsUDF.java]: The any other atomic type comment is repeated here and seems redundant. Exception is thrown for the non-atomic case below.

File: src/main/java/sparksoniq/spark/iterator/flowr/GroupByClauseSparkIterator.java [hunk 4]
```diff
@@ -280,81 +280,66 @@ private void linearizeTuples(List<FlworTuple> keyTuplePairs) {
...
+            DataType columnType = DataTypes.StringType;
+            typedFields.add(DataTypes.createStructField(columnName, columnType, true));
+            columnName = columnIndex + "-doubleField";
+            columnType = DataTypes.DoubleType;
             typedFields.add(DataTypes.createStructField(columnName, columnType, true));
         }
 
         df.sparkSession().udf().register("createGroupingColumns",
-                new GroupClauseCreateColumnsUDF(variableAccessExpressions, columnNames, typesForAllColumns),
+                new GroupClauseCreateColumnsUDF(variableAccessExpressions, columnNames, null),
```
Comment (CanBerker) [inline src/main/java/sparksoniq/spark/iterator/flowr/GroupByClauseSparkIterator.java]: I think this parameter can be removed from the UDF

File: src/main/java/sparksoniq/spark/iterator/flowr/GroupByClauseSparkIterator.java [hunk 5]
```diff
@@ -280,81 +280,66 @@ private void linearizeTuples(List<FlworTuple> keyTuplePairs) {
...
-
-        String udfSQL = DataFrameUtils.getSQL(inputSchema, -1, false);
-
-        Dataset<Row> columnTypesDf = df.sparkSession().sql(
-                String.format("select distinct(determineGroupingDataType(array(%s))) as `distinct-types` from input",
-                        udfSQL)
-        );
-        Object columnTypesObject = columnTypesDf.collect();
-        Row[] columnTypesOfRows = ((Row[]) columnTypesObject);
+//        df.sparkSession().table("input").cache();
```
Comment (CanBerker) [inline src/main/java/sparksoniq/spark/iterator/flowr/GroupByClauseSparkIterator.java]: Just as a reminder, let's not forget to remove the commented out code before merging.

**Pull request closed without merging**